src/vmm/src/device_manager/persist.rs

```diff
@@ -87,6 +103,21 @@ pub struct DeviceStates {
     pub net_devices: Vec<ConnectedNetState>,
     /// Vsock device state.
     pub vsock_device: Option<ConnectedVsockState>,
+    /// Balloon device state.
+    #[version(start = 2, ser_fn = "balloon_serialize")]
+    pub balloon_device: Option<ConnectedBalloonState>,
+}
+
+impl DeviceStates {
+    fn balloon_serialize(&mut self, target_version: u16) -> VersionizeResult<()> {
+        if target_version < 2 && self.balloon_device.is_some() {
+            return Err(VersionizeError::Semantic(
+                "Balloon is not compatible with the target version.".to_owned(),
```

Comment:
I think we should be more specific: "Target version does not implement the virtio-balloon device."

Reply:
Changed.